dd-library-php-setup.php

```diff
@@ -0,0 +1,833 @@
+    echo "    - the INI setting 'allow_url_fopen=1'\n";
+
+    exit(1);
+}
+
+/**
+ * Progress callback as specified by the ext-curl documentation.
+ *   see: https://www.php.net/manual/en/function.curl-setopt.php#:~:text=CURLOPT_PROGRESSFUNCTION
+ *
+ * @return int
+ */
+function on_download_progress($curlHandle, $download_size, $downloaded)
+{
+    global $progress_counter;
+
+    if ($download_size === 0) {
+        return 0;
+    }
+    $ratio = $downloaded / $download_size;
+    if ($ratio == 1) {
+        return 0;
+    }
+
+    // Max 20 dots to show progress
+    if ($ratio >= ($progress_counter + (1 / 20))) {
+        $progress_counter = $ratio;
+        echo ".";
+    }
+
+    flush();
+    return 0;
+}
+
+/**
+ * Extracts and normalizes a set of properties from PHP's ini values.
+ *
+ * @param string $binary
+ * @return array
+ */
+function ini_values($binary)
+{
+    $properties = [INI_CONF, EXTENSION_DIR, THREAD_SAFETY, PHP_API, IS_DEBUG];
+    $lines = [];
+    // Timezone is irrelevant to this script. This is a quick-and-dirty workaround to the PHP 5 warning with missing
+    // timezone
+    exec($binary . " -d date.timezone=UTC -i", $lines);
+    $found = [];
+    foreach ($lines as $line) {
+        $parts = explode('=>', $line);
+        if (count($parts) === 2 || count($parts) === 3) {
+            $key = trim($parts[0]);
+            if (in_array($key, $properties)) {
+                $found[$key] = trim(count($parts) === 2 ? $parts[1] : $parts[2]);
+            }
+        }
+    }
+    return $found;
+}
+
+function is_truthy($value)
+{
+    $normalized = trim(strtolower($value));
+    return in_array($normalized, ['1', 'true', 'yes', 'enabled']);
+}
+
+/**
+ * @param array $phpVersions
+ * @param string $prefix Default ''. Used for testing purposes only.
+ * @return array
+ */
+function search_php_binaries($prefix = '')
+{
+    $results = [];
+
+    // First, we search in $PATH, for php, php7, php74, php7.4, php7.4-fpm, etc....
+    foreach (build_known_command_names_matrix() as $command) {
+        $path = exec("command -v $command");
+        if ($resolvedPath = resolve_command_full_path($command)) {
+            $results[$command] = $resolvedPath;
+        }
+    }
+
+    // Then we search in known possible locations for popular installable paths on different systems.
+    $standardPaths = [
+        $prefix . '/usr/bin',
+        $prefix . '/usr/sbin',
+    ];
+    $remiSafePaths = array_map(function ($phpVersion) use ($prefix) {
+        list($major, $minor) = explode('.', $phpVersion);
+        return "${prefix}/opt/remi/php${major}${minor}/root/usr/sbin";
+    }, get_supported_php_versions());
+
+    foreach (($standardPaths + $remiSafePaths) as $knownPath) {
+        $pathsFound = [];
+        // phpcs:disable Generic.Files.LineLength.TooLong
+        exec(
+            "find -L $knownPath -type f -executable -regextype sed -regex '.*/php\(-fpm\)\?\([0-9][\.]\?[0-9]\?\)\?\(-fpm\)\?' 2>/dev/null",
```

Comment:
Valid concern, I want to see if I can easily `-or` first though, as for searches that might have to scan many directories, I prefer to use the native `find` rather than PHP which will be - anecdotally - slower.

Reply:
In the end I decided to refactor a bit this part of the code to address your valid concern, but at the same time to rely on native commands which should be faster. Let me know what you think about it